Feat: add update account setting page

SeijaHuang opened the pull request:
<!-- link to tickets -->
https://beequant.atlassian.net/jira/software/projects/CP/boards/1?selectedIssue=CP-113
---

## What happens here?
Add app setting page:
- Implemented language change
- Implemented theme change Add internationalization

Add next-intl library to achieve internationalization routing
- Modify middleware with next-intl middleware
- Modify next navigations with next-intl shared routing, including redirect, useRouter etc. Resolve CP-113

Screenshot:
1. Login Page
![10001724203600_ pic](https://github.com/user-attachments/assets/26f2733b-89b5-42f7-9c37-bfe2a994f9ac)
![10011724203612_ pic](https://github.com/user-attachments/assets/d98df234-32f4-441d-bae5-96d6fd2fd204)
![10031724203717_ pic](https://github.com/user-attachments/assets/bcf4ef1a-18aa-4525-b9cc-f76970dbd14c)
![10021724203700_ pic](https://github.com/user-attachments/assets/2c157194-af92-4eca-a0da-f1f274005aaa)

2. Rigester Page
![10041724203754_ pic](https://github.com/user-attachments/assets/3310286d-8c88-41e9-a9eb-22f57f2d983e)
![10051724203755_ pic](https://github.com/user-attachments/assets/31094d79-243c-4ae8-b253-eb60c264ea23)
![10061724203781_ pic](https://github.com/user-attachments/assets/98716b8c-8eae-4d00-bf06-75eccaab8ae1)
![10071724203781_ pic](https://github.com/user-attachments/assets/a0a9bd3a-eaaa-4c00-a904-a49ce3d4aa13)

4. Dashboard
![10091724203866_ pic](https://github.com/user-attachments/assets/881178f8-2f7c-4f65-9141-6aca04b08097)
![10081724203851_ pic](https://github.com/user-attachments/assets/85f2474b-9290-4318-8410-7980eb9cf907)

6. App Setting Page
![10101724204036_ pic](https://github.com/user-attachments/assets/0ccb7683-75f0-4cb1-8262-43c8eeb580e0)
![10111724204055_ pic](https://github.com/user-attachments/assets/4f540f1b-fe08-4fad-9b90-1f312d45c87d)

7. Account Setting Page
![10121724204158_ pic](https://github.com/user-attachments/assets/b0f567d0-5c01-414c-975f-f8cb05a75d2e)
![10131724204234_ pic](https://github.com/user-attachments/assets/af97f4d1-fdb0-41c8-aac8-5e65ac948e27)

8. Register Succeed Page
![10151724205221_ pic](https://github.com/user-attachments/assets/103072fc-1e1e-4c5c-ba99-75f2abd72e60)
![10141724205209_ pic](https://github.com/user-attachments/assets/4d528309-4bc2-4d21-adcc-a339449b8224)

## How do I know this is working?
in this branch,  you can use /en/pathname or /zh-cn/pathname to see all changes

## Any notes?

<!--
  New PR Checklist

- Have you tag the right reviewer?
- Have you dm the reviewer?
- Maybe @ the reviewer in channel just to be safe

- is CI passing in your pr?
- Have you update the ticket with the right status?
-->

NorrisWu0 commented on README.md:
```diff
@@ -1,16 +1,10 @@
-## How do I run this
-
-### Preping local database
+## What is this?
```
What's going on here?

SeijaHuang replied:
reset with dev branch

NorrisWu0 commented on apps/client/package.json:
```diff
@@ -2,7 +2,7 @@
   "version": "0.1.0",
   "scripts": {
-    "dev": "concurrently 'yarn:dev-*' --restart-tries 3 --restart-after 3000",
+    "dev": "yarn dev-codegen & yarn dev-next",
```
yeah, you need to fix these kind of changes before I can do reviews

SeijaHuang replied:
reset with dev branch

NorrisWu0 commented on apps/client/src/module/appsetting/app-setting-page.tsx:
```diff
@@ -0,0 +1,93 @@
+    switchThemeToOption,
+    switchOptionToTheme,
+    isTheme,
+  } = useThemeSwitcher();
```
there's already a existing theme switcher, check under `module/system`

NorrisWu0 commented on apps/client/src/module/appsetting/language-switcher.ts:
```diff
@@ -0,0 +1,47 @@
+    return localesSwitchRules[lang];
+  };
+
+  const isLocale = (value: string): value is locales => {
```
~so this looks like a util function? why is this here?~

we provides the options to switch languages yeah? why do you need to check if a value is a locale?

and why do you need `value is locales`?

SeijaHuang replied:
Changed using as

NorrisWu0 commented on apps/client/src/module/appsetting/language-switcher.ts:
```diff
@@ -0,0 +1,47 @@
+import { getNextLocale } from "@src/actions";
+import { useEffect, useState } from "react";
+
+type locales = "en" | "zh-cn";
```
You know what to do here

SeijaHuang replied:
yes, import form the config file, the only place stored optional locales

NorrisWu0 commented on apps/client/src/module/appsetting/language-switcher.ts:
```diff
@@ -0,0 +1,47 @@
+    });
+  }, []);
+
+  // Locales switch function and rules
```
ummmm, why do you need the switch functions and rules? we know what the locales are yeah?

SeijaHuang replied:
changed the data structure as you mentored

NorrisWu0 commented on apps/client/src/module/appsetting/app-setting-page.tsx:
```diff
@@ -0,0 +1,93 @@
+    theme: [t("Shared.darkTheme"), t("Shared.lightTheme")],
+  };
+
+  const { control } = useForm<FormValues>({
```
first up, why do you need to use `react-hook-form` here?

second, it doesn't seem to be used for anything?

SeijaHuang replied:
yes, there is no need here

NorrisWu0 commented on apps/client/src/module/common/select.tsx:
```diff
@@ -0,0 +1,39 @@
+        className="focus:shadow-accent-400 dark:focus:shadow-accent-400 shadow-primary-100 dark:shadow-primary-600 shadow-inset flex-1 bg-transparent px-2 py-2 text-sm transition-[box-shadow] duration-300 placeholder:text-sm focus:outline-none"
+        {...field}
+        {...props}
+        onChange={(e) => onChangeHandler(e)}
```
is this working? if you don't want react-hook-form to control this input, then you probably don't need the `useController` hook yeah?

SeijaHuang replied:
actually in the previous file, there is a useController hook

NorrisWu0 commented on apps/client/src/module/common/icon.tsx:
```diff
@@ -82,9 +82,7 @@ type LinearIconProps = { icon: LinearIcons; className?: string };
-      className={twMerge(
-        clsx(`lnr lnr-${icon} text-primary-200 text-base`, className),
-      )}
+      className={twMerge(clsx(`lnr lnr-${icon} text-primary-200`, className))}
```
oh, what's wrong with `text-base`?

SeijaHuang replied:
reset all these changes

NorrisWu0 commented on apps/client/src/module/common/controlled-input.tsx:
```diff
@@ -50,7 +50,7 @@ export function ControlledTextInput<TFieldValues extends object>({
   return (
     <label className={inputVariants({ direction })}>
-      <span>
+      <span className="text-[13px]">
```
we got `text-base` yeah? do we need arbitrary value here?

SeijaHuang replied:
reset all these changes

NorrisWu0 commented on apps/client/src/module/dashboard/layout/sidebar-menu.tsx:
```diff
@@ -45,8 +45,8 @@ function SidebarMenuItem(props: SidebarMenuItem) {
         <span
```
I'm gonna assume these style changes are fine

NorrisWu0 commented on apps/client/src/module/dashboard/account/setting/setting-form.tsx:
```diff
@@ -3,40 +3,37 @@ import { zodResolver } from "@hookform/resolvers/zod";
+export function AccountSettingForm() {
+  const t = useTranslations();
+
+  const formSchema = z.object({
```
assuming this schema is good

NorrisWu0 commented on apps/client/src/module/appsetting/theme-switcher.ts:
```diff
@@ -0,0 +1,53 @@
+"use client";
```
not gonna comment on this, since we already got one

NorrisWu0 commented on apps/client/src/module/appsetting/language-switcher.ts:
```diff
@@ -0,0 +1,47 @@
+  const switchLanguageValue = (
+    lang: keyof typeof localesSwitchRules,
+  ): string => {
+    return localesSwitchRules[lang];
```
so what does this function do?

SeijaHuang replied:
changed the data structure as you mentored

NorrisWu0 commented on apps/client/src/module/appsetting/app-setting-page.tsx:
```diff
@@ -0,0 +1,93 @@
+      </div>
+      {!loading && (
+        <form className="my-[30px] flex flex-col gap-y-5">
+          <Select
```
again, it doesn't seems like you need a controlled select here.

SeijaHuang replied:
yeah， already delete react form hook

NorrisWu0 commented on apps/client/src/middleware.ts:
```diff
@@ -1,16 +1,27 @@
 import { NextRequest, NextResponse } from "next/server";
```
assuming this is all good

NorrisWu0 commented on apps/client/src/i18n.ts:
```diff
@@ -0,0 +1,14 @@
+import { notFound } from "next/navigation";
```
assuming this is all good

NorrisWu0 reviewed (changes requested):
okay, well this is a lot, A LOT of changes. going to assume the files I've haven't look at are good (mostly files under app directory and static files),

SeijaHuang committed:
Feat: add internationalization
Add app setting page:
- Implemented language change
- Implemented theme change
Add internationalization
- Add next-intl library to achieve internationalization routing
- Modify middleware with next-intl middleware
- Modify next navigations with next-intl shared routing, including redirect, useRouter etc.
Resolve CP-113

Chris-Wang commented on apps/client/next.config.mjs:
```diff
@@ -1,6 +1,10 @@
+
+const withNextIntl = createNextIntlPlugin();
+
 /** @type {import('next').NextConfig} */
```
what's this comment for?

SeijaHuang replied:
I would delete it!

SeijaHuang replied:
This is form next doc, which is used to add some type checking in config fiels.
<img width="695" alt="image" src="https://github.com/user-attachments/assets/c87d5fc4-0dc2-4039-90cf-11e0a4344d42">

Chris-Wang commented on apps/client/src/actions/get-next-locale.ts:
```diff
@@ -0,0 +1,13 @@
+  if (!result) {
+    return defaultLocale;
+  }
+  const nextLocale = result;
```
why not directly return result?

SeijaHuang replied:
Because the result of NEXT_LOCALE maybe undefined(user never visited our page)
<img width="682" alt="image" src="https://github.com/user-attachments/assets/f1ec6374-913b-44bd-a519-74d254af6de5"> type undefined is not assignable to the locales('en' | 'zh-cn'), so if result is undefined it needs to return defaultLocale, which is 'en'

Chris-Wang replied:
at line 11, the result shouldn't be falsy, so what's the nextLocale for?

SeijaHuang replied:
The NEXT_LOCALE is set by next-intl. Once user have visited or changed language the language setting will be stored in the cookie.  In this case,  by getting the preferred language or default language from the cookie, the setting page select will render the responding language setting.

Chris-Wang replied:
... 

shouldn't we remove line 11 and 12, and just make it
` return result`?

Chris-Wang commented on apps/client/src/app/[locale]/(auth)/register/page.tsx:
```diff
@@ -1,4 +1,4 @@
-import { AuthRoute, RegisterForm } from "../../../module/auth";
+import { AuthRoute, RegisterForm } from "../../../../module/auth";
```
don't we have alias of module, we probably think of using @ for import

SeijaHuang replied:
refactored

Chris-Wang commented on apps/client/src/utils/user-help-messgage.ts:
```diff
@@ -1,5 +1,3 @@
 export enum UserHelpMsgs {
   DisplayNameSettingPage = "Display is optional: it should between 4 to 15 characters, only letters, numbers, dashes and underscore are allowed!",
```
is `DisplayNameSettingPage` still being used?

SeijaHuang replied:
No, I've deleted it already.

Chris-Wang replied:
if UserHelpMsgs is unused, why not remove it?

Chris-Wang commented on apps/client/src/module/system/system.ts:
```diff
@@ -0,0 +1,23 @@
+import { Locales } from "@src/configs/configs";
+import { create } from "zustand";
+
+export type Theme = "dark" | "light" | "system";
```
`system` for ?

SeijaHuang replied:
system setting includes language and theme setting, there two global states are system scope, so I gave the name system for them

Chris-Wang replied:
when could the `Theme` be `system`? 

I assume `Theme` represent colour theme of the system here, do we make it like color theme or language theme here?

SeijaHuang replied:
delete this system.ts, put language and theme settings back to their responding files

Chris-Wang commented on apps/client/src/module/system/theme-switcher.tsx:
```diff
@@ -110,7 +82,12 @@ export function useTheme() {
   }, [setTheme]);
 
-  return { theme, setTheme };
+  function themeSwitcher({ e }: { e: React.ChangeEvent<HTMLSelectElement> }) {
```
I like this refactoring

Chris-Wang commented on apps/client/src/middleware.ts:
```diff
@@ -1,16 +1,27 @@
 export async function middleware(request: NextRequest) {
   const token = request.cookies.get("token");
+  const pathname = request.nextUrl.pathname;
+  const pagesWithoutToken =
```
could we have an array to store those page names for better maintainability

SeijaHuang replied:
Sure!

Chris-Wang reviewed (changes requested):
better to split PR into smaller ones the next time, as we discussed in the standup

Chris-Wang reviewed (approved):
LGTM
